### DISCUSSION TITLE
How to create a policy in storage where each user can have only one image stored in the bucket?

### DISCUSSION THREAD

**Discussion by rshtj** [supabase/supabase#2844] [Questions] [answered]
I set up a project using [user management starter template](https://supabase.io/docs/guides/with-nextjs#bonus-profile-photos).
In this application, each user has the option to upload their profile picture, but their previous ones are still stored in the bucket.

Is there any way to do something like creating a policy or triggering a function and delete an old profile picture when a user updates to the new one which can save storage space?

**Comment by dshukertjr** [comment, marked as answer]
Hi @shanmukhrvl!

Different approach from your methods, but could making the user upload at same file path with their user ID? This way, they will always be overriding their existing images!

```typescript
const fileName = `${userId}.${fileExt}`
const { data } = supabase.storage.from('users').upload(fileName, file, { upsert: true })
```

**Reply by rshtj** [reply]
Thank you for the suggestion, the problem with this approach is it still stores older images if the user uploads with a different file extension. I think I should probably look into converting the uploaded image into a single format like jpeg and then store it in the bucket with userId as the filename

**Reply by dshukertjr** [reply]
@shanmukhrvl 
>the problem with this approach is it still stores older images if the user uploads with a different file extension

Yeah, that might be true, but as you said, usually whenever you let users upload an image, you would want to compress it, which would align the extension of the images, so did not touch on that within my answer!

**Reply by inian** [reply]
Tylers approach is the way to go here. You can even ignore the fileExtension in the file name and the server should be able to detect the content type if you are passing in a Blob or FormData.

**Reply by rshtj** [reply]
@inian 
> You can even ignore the fileExtension in the file name and the server should be able to detect the content type if you are passing in a Blob or FormData.

Tried this approach, but it throws an error ```new row violates row-level security policy (USING expression) for table "objects"``` 

It seems I am now having the same problem as #2679 . Updating insert and update permissions under ```storage.objects``` doesn't work either. Are there any other permissions that I need to add?

![supapolicy](https://user-images.githubusercontent.com/56866810/129870858-b25850fc-e0f2-453d-8814-e664f3c2d12d.png)
